### PR TITLE
feat(client): add support for cloud storage configuration per project

### LINKED ISSUE #2736: Support the new storage service
## Motivation

We need to support external cloud storages through the new service ([API reference](https://github.com/SwissDataScienceCenter/renku-data-services/blob/main/components/renku_data_services/storage_schemas/api.spec.yaml)).

:information_source:  _Mind that there is no way to deploy the storage service yet, but this will soon be possible._

:exclamation: do _not_ start working on this until the refactoring coming with the persistent sessions is finished (see #2626)

## Changes

There will be a rather big change in the flow for including storages in a session:

- Currently, users need to add the storage on the "Start session" page manually _each time_ they start a session.
    image

- With the new service, this will be split into 2 phases:
    - Link storages to a project.
    - Start a session using the storages, asking for the credentials whenever needed.

Mind that storages will be _per project_. Any maintainer can set up storages for a project, affecting the next sessions for each user.

## Proposal

We should consider the following changes:

- Handle the storages in a new Settings tab named `Storages`. There, users should see the already attached storages, if any, and add, delete, or edit them _if they are at least project maintainers_.
- Change the "Cloud Storage" section on the "Start session" page. We need to:
    - Query for the storages attached to the project.
    - Allow users to disable single storages; we will pass the list of non-disable storages to the `POST /servers` API. By default, all are ticked.
    - Ask for credentials for each storage requiring them. We cannot cache them somewhere in the backend for the time being, so not providing credentials for a private storage won't work. Users _must_ either provide credentials or disable it for the session.
- Check storages when using the quickstart session button; if any of them is private, interrupt the flow (as for errors) and show a warning mentioning the user needs to specify storages credentials.

On the "Start session" page we might want always to show a "Cloud Storage" section. If no storages are set up, we can mention something like "You can configure your project to automatically mount Cloud storages like AWS S3" and add a configure button linking to the new storages setting page. We could consider having a Modal there to configure storages immediately, but that's a nice-to-have and unnecessary.

P.S: Mind that, due to some renku-notebooks limitations, we should initially support _only_ S3 buckets. This constraint will be relaxed later. Mind also that the storage service will recognize the type of storage, so there will be no need to ask users for the storage type they want to use.

### PR DESCRIPTION
**Pull request by leafty**
Fixes #2736.

/deploy renku=leafty/update-cypress-tests-cloud-storage renku-gateway=master renku-notebooks=master renku-data-services=main extra-values=dataService.image.pullPolicy=Always,global.gitlab.url=gitlab.dev.renku.ch,notebooks.cloudstorage.s3.enabled=true,amalthea.extraChildResources[0].group=com.ie.ibm.hpsys,amalthea.extraChildResources[0].name=datasets,notebooks.cloudstorage.s3.installDatashim=false

### CODE REVIEW

**Comment by lorenzo-cavazzi** [comment]
I was wondering if we should use another sub-tab in the settings page just for the Cloud Storage instead of filling even more the sessions tab. That way, we won't need a modal to interact with single storages.

The proposed solution might work well but I feel there is a little too much content on the page. Perhaps we should consider adding white cards as in the other session page to separate the different parts better visually.

P.S. be sure to coordinate with @jdupre81 on this

**Comment by leafty** [comment]
Some WIP for adding a new storage configuration:
image
image

**Comment by leafty** [comment]
Screenshots of this new feature:

### Anonymous users
Screenshot from 2023-10-06 15-04-11

### User has no access
Screenshot from 2023-10-06 15-04-39

### Variants depending on the support from `renku-notebooks`
Screenshot from 2023-10-06 15-05-33
Screenshot from 2023-10-06 15-06-38
Screenshot from 2023-10-06 15-06-57

### Add a new cloud storage
Screenshot from 2023-10-06 15-08-17
Screenshot from 2023-10-06 15-07-34
Screenshot from 2023-10-06 15-08-27
Screenshot from 2023-10-06 15-08-46
Screenshot from 2023-10-06 15-22-52

### View cloud storage
Screenshot from 2023-10-06 15-11-10

### Edit cloud storage
Screenshot from 2023-10-06 15-12-14

### Cloud storage not supported
Screenshot from 2023-10-06 15-12-58

### Cloud storage credentials
Screenshot from 2023-10-06 15-13-18

### Quick start interrupted for credentials
Screenshot from 2023-10-06 15-13-30

### Temporary cloud storage
Screenshot from 2023-10-06 15-18-03
Screenshot from 2023-10-06 15-17-39

**Review by ciyer — CHANGES_REQUESTED** [review]
I have reviewed all but 4 files. I'll need a little more time for the last four files, but I wanted to provide these comments now.

I like the code and the appearance in general! I have some suggestions for refactorings in a few places.

File: client/src/features/session/sessions.types.ts [hunk 1]
```diff
@@ -108,3 +108,26 @@ export interface PatchSessionParams {
...
+  | {
+      configuration: {
+        type: "azureblob";
+        endpoint: string;
+        secret_access_key: string;
+      };
+      readonly: boolean;
+      source_path: string;
+      target_path: string;
+    };
```
[the remark this answers is not in the document]
Comment (leafty) [reply client/src/features/session/sessions.types.ts]: Updated in [53ea0b9](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/53ea0b9574dcaf5003f7058243df533b012773f4)

File: client/src/features/session/hooks/options/useDefaultCloudStorageOption.hook.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+              sensitive_fields: sensitive_fields.map(({ name, ...rest }) => ({
+                ...rest,
+                name,
+                value: "",
+              })),
+            }
+          : {}),
+        ...storage,
+      })
+    );
```
[the remark this answers is not in the document]
Comment (leafty) [reply client/src/features/session/hooks/options/useDefaultCloudStorageOption.hook.ts]: I don't want to lift this into a module function. Moved it to a function on the same file in [53ea0b9](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/53ea0b9574dcaf5003f7058243df533b012773f4).

File: client/src/features/project/utils/projectCloudStorage.utils.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+export function getCredentialFieldDefinitions(
+  storageDefinition: CloudStorage
+): CloudStorageCredential[] | undefined {
+  const { sensitive_fields, storage } = storageDefinition;
+  const { configuration } = storage;
+
+  const providedSensitiveFields = Object.entries(configuration)
+    .filter(([, value]) => value === CLOUD_STORAGE_SENSITIVE_FIELD_TOKEN)
+    .map(([key]) => key);
```
[the remark this answers is not in the document]
Comment (leafty) [reply client/src/features/project/utils/projectCloudStorage.utils.ts]: Done in [53ea0b9](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/53ea0b9574dcaf5003f7058243df533b012773f4)

**Review by ciyer — CHANGES_REQUESTED** [review]
Thanks for making those changes. I am testing the feature and I have some feedback on the UI.

Adding cloud storage to the project settings is quite intuitive to use. One small suggestion, could you add some info text explaining that the `name` is used to determine the mount point? 

image

I wanted to add temporary storage to a session, but I have absolutely no idea what I need to enter into these fields:
image

Is this documented somewhere?

**Review by ciyer — CHANGES_REQUESTED** [review]
I tested with with public and private AWS buckets, and when everything is right, it works very nicely. 🎉 

When there are problem, though, it is basically impossible to debug, since there is no useful feedback available in the UI. I realize that this may be out of scope for this issue, but I just wanted to provide the feedback, so that this could be improved on in the future.

One thing I noticed about private S3 buckets -- the UI autofilled the `access_key_id` and the `secret_access_key`. For the former, I think this is probably ok, but I would have expected the `secret_access_key` to be treated like a password and 1. not shown in plaintext and 2. not autofilled without me authorizing the action.

**Comment by lokijuhy** [comment]
@leafty Re text copy in the screenshots:

I put suggestions for modified text in **bold**, and text removal in ~~strikethrough~~.

### Screenshot "Anonymous users"
  - "Only authenticated users can access cloud storage setting**s**."
  - "Log in to **view** **cloud storage** settings"

### Screenshots for Variants depending on the support from `renku-notebooks`
  - for all:
      - “… when starting local sessions from the CLI”. ← how about linking to the docs here? [renku session docs link](https://renku.readthedocs.io/en/latest/renku-python/docs/reference/commands/session.html) (or if we’re getting more specific docs for cloud storage, use those instead!)
  - Variant # 1:
      - "~~Note that~~ This instance of **RenkuLab** does not support…"
  - Variants # 2 & # 3:
      - “You can always mount cloud storage when starting local sessions with the CLI” - Do you mean “You can mount *other types of* cloud storage when starting local sessions with the CLI”?

### Screenshot for Temporary Cloud Storage dialog box
  - Add to the top of the box: “This cloud storage will be configured for this session only. To configure cloud storage permanently for this project, go to Project Settings.”

**Comment by leafty** [comment]
> One thing I noticed about private S3 buckets -- the UI autofilled the `access_key_id` and the `secret_access_key`. For the former, I think this is probably ok, but I would have expected the `secret_access_key` to be treated like a password and 1. not shown in plaintext and 2. not autofilled without me authorizing the action.

I think I would like to address this in a subsequent PR. We have no input of type "password" at the moment and making a nice one is not trivial.

File: client/src/features/project/components/AddCloudStorageButton.tsx [hunk 4]
```diff
@@ -0,0 +1,862 @@
...
+        <Form
+          className="form-rk-green"
+          noValidate
+          onSubmit={handleSubmit(onSubmit)}
+        >
+          {result.error && <RtkErrorAlert error={result.error} />}
+
+          <div className="form-rk-green">
+            <div className="mb-3">
+              <Label className="form-label" for="addCloudStorageName">
```
[the remark this answers is not in the document]
Comment (leafty) [reply client/src/features/project/components/AddCloudStorageButton.tsx:541]: Done in [b066797](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/b0667978e9aa57fae3043099379e8764be606f5f).

File: client/src/features/project/components/AddCloudStorageButton.tsx [hunk 5]
```diff
@@ -0,0 +1,862 @@
...
+          {result.error && <RtkErrorAlert error={result.error} />}
+
+          <div className="form-rk-green">
+            <p className="mb-3">
+              Advanded mode uses <code>rclone</code> configurations to set up
+              cloud storage.
+            </p>
+
+            <div className="mb-3">
+              <Label className="form-label" for="addCloudStorageName">
```
[the remark this answers is not in the document]
Comment (leafty) [reply client/src/features/project/components/AddCloudStorageButton.tsx:269]: Done in Done in [b066797](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/b0667978e9aa57fae3043099379e8764be606f5f).

File: client/src/features/project/components/AddCloudStorageButton.tsx [hunk 6]
```diff
@@ -0,0 +1,862 @@
...
+} from "../utils/projectCloudStorage.utils";
+
+import LazyRenkuMarkdown from "../../../components/markdown/LazyRenkuMarkdown";
+import { useGetNotebooksVersionsQuery } from "../../versions/versionsApi";
+import styles from "./AddCloudStorageButton.module.scss";
+
+export default function AddCloudStorageButton() {
+  const [isOpen, setIsOpen] = useState(false);
+  const toggle = useCallback(() => {
+    setIsOpen((open) => !open);
```
[the remark this answers is not in the document]
Comment (leafty) [reply client/src/features/project/components/AddCloudStorageButton.tsx:68]: Done in Done in [b066797](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/b0667978e9aa57fae3043099379e8764be606f5f).
Made the modal backdrop static, i.e. the modal can only be closed by clicking on the close button or icon.

**Comment by leafty** [comment]
## Screenshot "Anonymous users"

Updated:
image

## Screenshots for Variants depending on the support from `renku-notebooks`

Message about local sessions removed, as the CLI cannot start local sessions (Docker) with cloud storage.
image
image
image

## Screenshot for Temporary Cloud Storage dialog box

Added info box and explainer text about `rclone`.
image

**Review by andre-code — CHANGES_REQUESTED** [review]
Nice addition! I didn't conduct an exhaustive code review, as my colleagues have already done so, but I'd like to offer a couple of small suggestions to enhance the user experience:

1. Consider include an alert within the project session settings to inform users when they lack the necessary permissions, similar to how we handle it in the general project settings.

Screenshot at Oct 11 17-08-59

2. It might be beneficial to relocate the description for "rclone" to the input help label or position it below the "rclone" textarea for improved clarity.

Screenshot at Oct 11 16-57-49

**Review by andre-code — APPROVED** [review]
🚀

**Pull request merged**